tests/0113-cooperative_rebalance.cpp

```diff
@@ -26,311 +26,1714 @@
+
+    test_conf_init(&conf, NULL, 30);
+    test_conf_set(conf, "bootstrap.servers", bootstraps);
+    test_conf_set(conf, "security.protocol", "PLAINTEXT");
+    test_conf_set(conf, "group.id", groupid);
+    test_conf_set(conf, "auto.offset.reset", "earliest");
+    test_conf_set(conf, "enable.auto.commit", "false");
+    test_conf_set(conf, "partition.assignment.strategy", "cooperative-sticky");
+
+    c = test_create_consumer(groupid, rebalance_cb, conf, NULL);
+
+    test_consumer_subscribe(c, topic);
+
+    expect_rebalance("initial assignment", c,
+                     RD_KAFKA_RESP_ERR__ASSIGN_PARTITIONS,
+                     rd_false/*don't expect lost*/, 5+2);
+
+    /* Fail heartbeats */
+    rd_kafka_mock_push_request_errors(
+      mcluster, RD_KAFKAP_OffsetCommit,
+      5,
+      RD_KAFKA_RESP_ERR_ILLEGAL_GENERATION,
+      RD_KAFKA_RESP_ERR_ILLEGAL_GENERATION,
+      RD_KAFKA_RESP_ERR_ILLEGAL_GENERATION,
+      RD_KAFKA_RESP_ERR_ILLEGAL_GENERATION,
+      RD_KAFKA_RESP_ERR_ILLEGAL_GENERATION);
+
+    rd_kafka_commit(c, NULL, rd_false);
+
+    expect_rebalance("lost partitions", c,
+                     RD_KAFKA_RESP_ERR__REVOKE_PARTITIONS,
+                     rd_true/*expect lost*/, 10+2);
+
+    expect_rebalance("rejoin group", c,
+                     RD_KAFKA_RESP_ERR__ASSIGN_PARTITIONS,
+                     rd_false/*expect lost*/, 20+2);
+
+    TEST_SAY("Closing consumer\n");
+    test_consumer_close(c);
+
+    TEST_SAY("Destroying consumer\n");
+    rd_kafka_destroy(c);
+
+    TEST_SAY("Destroying mock cluster\n");
+    test_mock_cluster_destroy(mcluster);
+  }
+
+
+
+  int main_0113_cooperative_rebalance (int argc, char **argv) {
+    int i;
+
+    a_assign_tests();
+    b_subscribe_with_cb_test(true/*close consumer*/);
+
+    if (test_quick) {
+      Test::Say("Skipping tests c -> s due to quick mode\n");
+      goto end;
+    }
+
+    b_subscribe_with_cb_test(false/*don't close consumer*/);
+    c_subscribe_no_cb_test(true/*close consumer*/);
+    c_subscribe_no_cb_test(false/*don't close consumer*/);
+    d_change_subscription_add_topic(true/*close consumer*/);
+    d_change_subscription_add_topic(false/*don't close consumer*/);
+    e_change_subscription_remove_topic(true/*close consumer*/);
+    e_change_subscription_remove_topic(false/*don't close consumer*/);
+    f_assign_call_cooperative();
+    g_incremental_assign_call_eager();
+    h_delete_topic();
+    i_delete_topic_2();
+    j_delete_topic_no_rb_callback();
+    k_add_partition();
+    l_unsubscribe();
+    m_unsubscribe_2();
+    n_wildcard();
+    o_java_interop();
+    p_lost_partitions_heartbeat_illegal_generation_test();
+    q_lost_partitions_illegal_generation_test(rd_false/*joingroup*/);
+    q_lost_partitions_illegal_generation_test(rd_true/*syncgroup*/);
+    r_lost_partitions_commit_illegal_generation_test();
+    for (i = 1 ; i <= 6 ; i++) /* iterate over 6 different test variations */
+      s_subscribe_when_rebalancing(i);
 
+end:
```

Comment:
Are there any negative testing, like consumers with unmatched assignment strategies?

Reply:
i test that you can't call assign when in COOPERATIVE rebalancing mode and vice versa. i'll try to think up some more now.